File: control-plane/sdk/typescript/src/transport.ts
/**
 * Consumer SDK transports (TypeScript) — issue #41.
 *
 * The SDK talks to the platform through small seams so every client is fully
 * testable offline against doubles and a deployment swaps in real HTTP:
 *
 * - `Transport` — one request/response round trip returning the parsed JSON
 *   body (gateway single dispatch + control-plane client);
 * - `StreamTransport` — one streaming round trip yielding parsed JSON bodies,
 *   mirroring the gateway streaming handler (issue #16: incremental
 *   `{"event": ...}` envelopes then the terminal `{status, result, record}`).
 *
 * `HttpTransport` is a dependency-free fetch-based HTTP implementation of
 * `Transport`.  The streaming adapter (an SSE relay) is the documented
 * production wiring point; offline doubles implement `StreamTransport`
 * directly.  No SDK test touches the network.
 */

import { TransportError } from "./errors.js";

export interface RequestOptions {
  /** JSON-serializable request body (POST/PUT). */
  body?: unknown;
  /** Query parameters (undefined values are dropped). */
  query?: Record<string, string | number | boolean | undefined>;
  /** The short-lived session token (attached as `Authorization: Bearer`). */
  token?: string;
  /** Extra headers. */
  headers?: Record<string, string>;
}

export interface Transport {
  request(method: string, path: string, opts?: RequestOptions): Promise<Record<string, unknown>>;
}

export interface StreamTransport {
  requestStream(
    method: string,
    path: string,
    opts?: RequestOptions
  ): AsyncIterable<Record<string, unknown>>;
}

function dropUndefined(query: Record<string, string | number | boolean | undefined> | undefined) {
  if (!query) {
    return undefined;
  }
  const out: Record<string, string> = {};
  for (const [key, value] of Object.entries(query)) {
    if (value !== undefined) {
      out[key] = String(value);
    }
  }
  return Object.keys(out).length > 0 ? out : undefined;
}

function buildUrl(baseUrl: string, path: string, query?: Record<string, string>): string {
  let url = `${baseUrl.replace(/\/+$/, "")}${path}`;
  if (query) {
    url = `${url}?${new URLSearchParams(query).toString()}`;
  }
  return url;
}

/** Dependency-free HTTP transport (`fetch`) — the default production seam. */
export class HttpTransport implements Transport {
  private readonly baseUrl: string;
  private readonly timeoutMs: number;

  constructor(baseUrl: string, opts: { timeoutMs?: number } = {}) {
    if (!baseUrl || !/^https?:\/\//.test(baseUrl)) {
      throw new Error("base_url must be an http(s) URL");
    }
    this.baseUrl = baseUrl;
    this.timeoutMs = opts.timeoutMs ?? 30_000;
  }

  async request(method: string, path: string, opts: RequestOptions = {}): Promise<Record<string, unknown>> {
    const headers: Record<string, string> = { ...(opts.headers ?? {}) };
    if (opts.token) {
      headers["Authorization"] = `Bearer ${opts.token}`;
    }
    let body: BodyInit | undefined;
    if (opts.body !== undefined) {
      headers["Content-Type"] = "application/json";
      body = JSON.stringify(opts.body);
    }
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await fetch(buildUrl(this.baseUrl, path, dropUndefined(opts.query)), {
        method,
        headers,
        body,
        signal: controller.signal,
      });
    } catch (error) {
      throw new TransportError(`transport failure: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      clearTimeout(timer);
    }
    const text = await response.text();
    let parsed: unknown;
    try {
      parsed = text.length > 0 ? JSON.parse(text) : {};
    } catch {
      throw new TransportError("response was not valid JSON");
    }
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
      throw new TransportError("response body was not a JSON object");
    }
    return parsed as Record<string, unknown>;
  }
}

File: control-plane/sdk/python/aosdk/transport.py
"""Consumer SDK transports (issue #41).

The SDK talks to the platform through two small seams so every client is fully
testable offline against doubles and a deployment swaps in real HTTP:

- :class:`Transport` — one request/response round trip returning the parsed
  JSON body (used by the gateway single-dispatch and the control-plane
  client).  A real HTTP transport attaches the bearer token as the
  ``Authorization`` header.
- :class:`StreamTransport` — one streaming round trip yielding a sequence of
  parsed JSON bodies, mirroring the gateway streaming handler (issue #16:
  incremental ``{"event": {...}}`` envelopes then the terminal
  ``{"status", "result", "record"}`` envelope).  A production HTTP adapter
  relays the SSE chunks; the offline doubles yield the same sequence.

:class:`HttpTransport` is a dependency-free (stdlib ``urllib``) HTTP
implementation of :class:`Transport` used as the production default; the
streaming adapter is the documented production wiring point (an SSE relay).
No network call is ever made by the SDK tests.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Dict, Iterator, Mapping, Optional, Protocol


class Transport(Protocol):
    """One request/response round trip returning the parsed JSON body."""

    def request(
        self,
        method: str,
        path: str,
        *,
        body: Optional[Dict[str, Any]] = None,
        query: Optional[Dict[str, Any]] = None,
        token: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]: ...


class StreamTransport(Protocol):
    """One streaming round trip yielding parsed JSON bodies."""

    def request_stream(
        self,
        method: str,
        path: str,
        *,
        body: Optional[Dict[str, Any]] = None,
        query: Optional[Dict[str, Any]] = None,
        token: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Iterator[Dict[str, Any]]: ...


def _drop_none(query: Optional[Mapping[str, Any]]) -> Dict[str, Any]:
    return {k: v for k, v in (query or {}).items() if v is not None}


class HttpTransport:
    """Dependency-free HTTP transport (stdlib ``urllib``) — the default seam.

    Attaches the bearer token when supplied, serializes JSON bodies, decodes
    JSON responses and maps HTTP/network failures to typed SDK errors.  For
    streaming, implement :class:`StreamTransport` over your SSE adapter (the
    offline doubles already speak the same chunk contract).
    """

    def __init__(self, base_url: str, *, timeout: float = 30.0) -> None:
        if not base_url or not base_url.startswith("http"):
            raise ValueError("base_url must be an http(s) URL")
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    def _url(self, path: str, query: Optional[Dict[str, Any]]) -> str:
        url = f"{self.base_url}{path}"
        if query:
            from urllib.parse import urlencode

            url = f"{url}?{urlencode(_drop_none(query))}"
        return url

    def request(
        self,
        method: str,
        path: str,
        *,
        body: Optional[Dict[str, Any]] = None,
        query: Optional[Dict[str, Any]] = None,
        token: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        from .errors import ApiError, TransportError

        req_headers = dict(headers or {})
        if token:
            req_headers["Authorization"] = f"Bearer {token}"
        payload = None
        if body is not None:
            req_headers["Content-Type"] = "application/json"
            payload = json.dumps(body).encode("utf-8")
        request = urllib.request.Request(
            self._url(path, query), data=payload, method=method, headers=req_headers
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:  # noqa: S310
                raw = response.read().decode("utf-8")
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode("utf-8")
            try:
                parsed = json.loads(raw)
                if isinstance(parsed, dict):
                    return parsed
            except Exception:  # noqa: BLE001 - non-JSON error body
                pass
            raise ApiError(exc.code, "http_error", exc.reason) from exc
        except urllib.error.URLError as exc:
            raise TransportError(f"transport failure: {exc.reason}") from exc
        try:
            parsed = json.loads(raw)
        except Exception as exc:  # noqa: BLE001
            raise TransportError("response was not valid JSON") from exc
        if not isinstance(parsed, dict):
            raise TransportError("response body was not a JSON object")
        return parsed
